src/nested_pandas/nestedframe/core.py

```diff
@@ -518,18 +519,37 @@ def query(self, expr: str, *, inplace: bool = False, **kwargs) -> NestedFrame |
         # to the nest and repack.  Otherwise, apply it to this instance as usual,
         # since it operated on the base attributes.
         if isinstance(result, _SeriesFromNest):
+            if not is_bool_dtype(result.dtype):
+                raise ValueError("Query condition must evaluate to a boolean Series")
+
             nest_name, flat_nest = result.nest_name, result.flat_nest
-            new_flat_nest = flat_nest.loc[result]
-            result = self.copy()
-            result[nest_name] = pack_sorted_df_into_struct(new_flat_nest)
+
+            # Reset index to "ordinal" like [0, 0, 0, 1, 1, 2, 2, 2]
+            flat_nest = flat_nest.set_index(self[nest_name].array.list_index)
+            query_result = result.set_axis(self[nest_name].array.list_index)
```

Comment:
This property `self[nest_name].array.list_index` does involve a (small) amount of computation; should it be pulled into a variable in order to be done once?  Or is the goal here to ensure that there are two distinct objects?

Reply:
We can make it a method and pull to a variable. Could you please do it if you have some time?